src/Language/PureScript/TypeChecker/Roles.hs

```diff
@@ -109,31 +130,37 @@ inferRoles env tyName
             -- * If the role is phantom, terminate, since the argument's use of
             --   our parameters is unimportant.
             TypeConstructor _ t1Name ->
-              let t1Roles = inferRoles env t1Name
+              let roles = M.lookup t1Name roleEnv'
+                  inferred = fromMaybe (repeat Phantom) roles
```

Comment:
In a couple of places you use `Phantom` as a default for a lookup. Can you clarify why this is desirable?

Reply:
The phantom role is used as a default because it is the least restrictive, so later inference passes can only refine it to a more precise role. We always prefer the more restrictive role inferred for the same parameter in the `Semigroup` instance for `RoleMap` and `updateRoles`.

For instance, given these definitions:

```purs
data D1 a = D1 a
data D2 a = D2 (D1 a)
```

When inferring `D2` roles we discover they depend on those of `D1`, but we don’t know them yet so we pick phantom for `a` and store that in our role environment. Then we infer the type parameter of `D1` is representational, store that in our role environment and go back to `D2`. This time we know `D1` roles so we can infer `a` is representational and update the role environment by taking the more restrictive of phantom and representational for `a`, which is representational.